File: assets/reviews.js
/*
Copyright 2016 Google Inc. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

var gitAppraiseWeb=angular.module("gitAppraiseWeb", []);

// Get a repository name from the full path.
function getLastPathElement(path) {
  var slashIndex = path.lastIndexOf("/");
  if (slashIndex > 0) {
    return path.substring(slashIndex+1, path.length);
  }
  return path;
}

// Get a review summary from the full description.
function getSummary(desc) {
  var result = desc;
  var newlineIndex = desc.indexOf("\n");
  if (newlineIndex > 0) {
    result = desc.substring(0, newlineIndex);
  }
  if (result.length > 80) {
    result = result.substring(0, 80);
  }
  return result;
}

gitAppraiseWeb.controller("listRepos", function($scope,$http) {
  $http.get("/api/repos").success(
    function(response) {$scope.repositories = processListReposResponse(response);});

  function processListReposResponse(response) {
    var repos = [];
    for (var i in response) {
      var path = response[i].path;
      repos.push(new Repo(response[i].id, getLastPathElement(path)));
    }
    return repos;
  }

  function Repo(id, name) {
    this.id = id;
    this.name = name;
  }
});

gitAppraiseWeb.controller("listReviews", function($scope,$http,$location) {
  var repo = $location.search()['repo'];
  $scope.repo = repo;
  $http.get("/api/repo_summary?repo=" + repo).success(
    function(response) {$scope.path = getLastPathElement(response.path);});
  listAllReviews("/api/open_reviews?repo=" + repo, function(response){
    $scope.openReviews = response;
  });
  listAllReviews("/api/closed_reviews?repo=" + repo, function(response){
    $scope.closedReviews = response;
  });

  function addPageItems(page, reviews) {
    for (var i in page.items) {
      var revision = page.items[i].revision;
      var timestamp = page.items[i].request.timestamp;
      var desc = page.items[i].request.description;
      reviews.push(new Review(revision, timestamp, desc, getSummary(desc)));
    }
  }

  function listReviewsResponseProcessor(baseQuery, callback) {
    var reviews = [];
    function processor(response) {
      addPageItems(response, reviews);
      if (response.nextPageToken) {
        $http.get(baseQuery + "&page=" + response.nextPageToken).success(processor);
      } else {
        callback(reviews);
      }
    }
    return processor;
  }

  function listAllReviews(baseQuery, callback) {
    $http.get(baseQuery).success(listReviewsResponseProcessor(baseQuery, callback));
  }

  function Review(revision, timestamp, desc, summary) {
    this.revision = revision;
    this.timestamp = timestamp;
    this.desc = desc;
    this.summary = summary;
  }
});

gitAppraiseWeb.controller("getReview", function($scope,$http,$location) {
  var repo = $location.search()['repo'];
  var review = $location.search()['review'];
  $http.get("/api/repo_summary?repo=" + repo).success(
    function(response) {$scope.path = getLastPathElement(response.path);});
  $http.get("/api/review_details?repo=" + repo + "&review=" + review).success(
    function(response) {
      $scope.details = response;
      loadSnippets(response.comments);
    });
  $http.get("/api/review_diff?repo=" + repo + "&review=" + review).success(
    function(response) {$scope.diff = response;});

  function loadSnippets(commentThreads) {
    var commentLocations = {};
    for (var i in commentThreads) {
      var commentThread = commentThreads[i];
      if ('location' in commentThread.comment) {
        var location = commentThread.comment.location;
        if (('commit' in location) && ('path' in location)) {
          var commit = location.commit;
          var path = location.path;
          if (!(commit in commentLocations)) {
            commentLocations[commit] = {};
          }
          var commitPaths = commentLocations[commit];
          if (!(path in commitPaths)) {
            commitPaths[path] = {};
          }
          var pathLines = commitPaths[path];
          if ('range' in location) {
            var range = location.range;
            if ('startLine' in range) {
              var line = range.startLine;
              if (!(line in pathLines)) {
                pathLines[line] = [];
              }
              var lineThreads = pathLines[line];
              lineThreads.push(commentThread.hash);
            }
          }
        }
      }
    }

    for (var commit in commentLocations) {
      var commitPaths = commentLocations[commit];
      for (var path in commitPaths) {
        var pathLines = commitPaths[path];
        var reader = snippetReader(path, commit, pathLines, commentThreads);
        $http.get("/api/repo_contents?repo=" + repo + "&commit=" + commit + "&file=" + path).success(reader);
      }
    }
  }

  function snippetReader(path, commit, pathLines, commentThreads) {
    return function(response) {
      var contentLines = response.split("\n");
      for (var line in pathLines) {
        if (line > 0 && line <= contentLines.length) {
          var startingLine = Math.max(0, line - 5);
          var endingLine = Math.max(0, line - 1);
          var snippetLines = [];
          for (var i = startingLine; i <= endingLine; i++) {
            snippetLines.push(new SnippetLine(i+1, contentLines[i]));
          }
          var snippet = new Snippet(commit, path, snippetLines);
          for (var h in pathLines[line]) {
            var hash = pathLines[line][h];
            for (var i in commentThreads) {
              var commentThread = commentThreads[i];
              if (commentThread.hash == hash) {
                commentThread.snippet = snippet;
              }
            }
          }
        }
      }
    };
  }

  function SnippetLine(lineNumber, contents) {
    this.lineNumber = lineNumber;
    this.contents = contents;
  }

  function Snippet(commit, path, lines) {
    this.commit = commit;
    this.friendlyCommit = commit.substring(0,6);
    this.path = path;
    this.lines = lines;
  }
});

File: api/reviews.go
/*
Copyright 2016 Google Inc. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package api

import (
	"github.com/google/git-appraise/repository"
	"github.com/google/git-appraise/review"
)

// CommitOverview encapsulates the fine-grained details of a commit.
type CommitOverview struct {
	ID      string                    `json:"id"`
	Details *repository.CommitDetails `json:"details"`
}

// DiffSummary summarizes one of the diffs included in a review.
//
// This summary includes the list of all commits that can be used to construct such a diff.
type DiffSummary struct {
	ReviewCommits []CommitOverview `json:"reviewCommits,omitEmpty"`
	LeftHandSide  string           `json:"leftHandSide"`
	RightHandSide string           `json:"rightHandSide"`
	Contents      string           `json:"contents"`
}

// ReviewListResponse represents a single `page` in a list of reviews.
type ReviewListResponse struct {
	Items         []review.Summary `json:"items"`
	NextPageToken string           `json:"nextPageToken,omitEmpty"`
}

func paginateReviews(reviews []review.Summary, maxPerPage int) [][]review.Summary {
	var results [][]review.Summary
	var currID int
	var currPage []review.Summary
	for _, r := range reviews {
		if currID >= maxPerPage {
			results = append(results, currPage)
			currPage = nil
			currID = 0
		}
		currPage = append(currPage, r)
		currID++
	}
	results = append(results, currPage)
	return results
}

// getReviewBase gets the earliest commit that can be used as a left-hand-side
// when generating diffs for a review.
//
// This is different from the `GetBaseCommit` method because that one only
// focuses on the current diff, whereas this method also includes historical diffs.
func getReviewBase(r *review.Review) (string, error) {
	if r.Request.BaseCommit != "" {
		return r.Request.BaseCommit, nil
	}
	submittedBase, err := r.Repo.GetLastParent(r.Revision)
	if err != nil {
		return "", err
	}
	if r.Submitted {
		return submittedBase, nil
	}
	targetRefBase, err := r.Repo.ResolveRefCommit(r.Request.TargetRef)
	if err != nil {
		return "", err
	}
	return r.Repo.MergeBase(targetRefBase, submittedBase)
}

// NewDiffSummary constructs a new instance of DiffSummary.
//
// If the `lhs` or `rhs` arguments are empty or out of bounds, then the
// review's base or head commits are used instead.
func NewDiffSummary(reviewDetails *review.Review, lhs, rhs string) (*DiffSummary, error) {
	base, err := getReviewBase(reviewDetails)
	if err != nil {
		return nil, err
	}
	head, err := reviewDetails.GetHeadCommit()
	if err != nil {
		return nil, err
	}
	reviewCommits := []string{base}
	subsequentCommits, err := reviewDetails.Repo.ListCommitsBetween(base, head)
	if err != nil {
		return nil, err
	}
	reviewCommits = append(reviewCommits, subsequentCommits...)
	commitsMap := make(map[string]interface{})
	var commitOverviews []CommitOverview
	for _, commit := range reviewCommits {
		commitsMap[commit] = nil
		details, err := reviewDetails.Repo.GetCommitDetails(commit)
		if err != nil {
			return nil, err
		}
		commitOverviews = append(commitOverviews, CommitOverview{
			ID:      commit,
			Details: details,
		})
	}
	if _, ok := commitsMap[lhs]; !ok {
		lhs, err = reviewDetails.GetBaseCommit()
		if err != nil {
			return nil, err
		}
	}
	if _, ok := commitsMap[rhs]; !ok {
		rhs = head
	}
	diff, err := reviewDetails.Repo.Diff(lhs, rhs)
	if err != nil {
		return nil, err
	}
	return &DiffSummary{
		ReviewCommits: commitOverviews,
		LeftHandSide:  lhs,
		RightHandSide: rhs,
		Contents:      diff,
	}, nil
}
